File: JS/Tree/450. Delete Node in a BST.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} key
 * @return {TreeNode}
 */
var deleteNode = function(root, key) {
    let current = root;
    let previous;
    while(current) {
        if (current.val === key) {
            let newNode;
            if (current.left) {
                let temp = current.left;
                while(temp.right) {
                    temp = temp.right;
                }
                temp.right = current.right;
                newNode = current.left;
            } else {
                newNode = current.right;
            }
            if (previous) {
                if (previous.val > key) {
                    previous.left = newNode;
                } else {
                    previous.right = newNode;
                }
                break;
            } else {
                return newNode;
            }
        } else if (current.val > key) {
            previous = current;
            current = current.left;
        } else {
            previous = current;
            current = current.right;
        }
    }
    return root;
};

File: Golang/Tree/450.Delete-node-in-a-bst.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func deleteNode(root *TreeNode, key int) *TreeNode {
	if root != nil {
		if root.Val == key {
			if root.Right == nil {
				return root.Left
			}
			node := root.Right
			for node != nil && node.Left != nil {
				node = node.Left
			}
			node.Left = root.Left
			return root.Right
		} else if root.Val < key {
			root.Right = deleteNode(root.Right, key)
		} else {
			root.Left = deleteNode(root.Left, key)
		}
	}
	return root
}
